packages/cli/internal/pkg/aws/cdk/execute_cdk.go

```diff
@@ -82,12 +93,59 @@ func processOutputs(stdout *bufio.Scanner, stderr *bufio.Scanner, executionName
 	}
 	go func() {
 		defer wait.Done()
+		// We can't just scan through lines, because the MFA prompt is a
+		// partial line and we need to see it. So scan runes instead and do our
+		// own line buffering.
```

Comment:
Are there tests we can write/update for this behavior?

Reply:
There are no existing tests for `processOutputs` that I can update.

There are tests [like this](https://github.com/aws/amazon-genomics-cli/blob/a5338edad6e9b4cc9eebc0ed938e956cce15a17b/packages/cli/internal/pkg/aws/cdk/execute_cdk_test.go#L84-L97) for `executeCdkCommandAndCleanupDirectory`. It's a little hard for me to put together why they're doing any of the things they're doing; they don't appear to be documented.

I think what's happening is the tests use fake CDK child process, which is actually [this function](https://github.com/aws/amazon-genomics-cli/blob/a5338edad6e9b4cc9eebc0ed938e956cce15a17b/packages/cli/internal/pkg/aws/cdk/execute_cdk_test.go#L133), and tell it to do one of a few things by passing it some particular arguments. Then they check to see if e.g. a failing exit code was properly detected, or if a line of output was properly parsed.

To use this approach to test the MFA prompt detection and parsing, and the proper forwarding of user input, we would have to:
1. Extend the child process to have a mode where it prints the MFA prompt and tries to read an MFA code. Easy.
2. Get the child process to make sure that the *right* MFA code was received, and somehow get the parent test to fail if it wasn't. I *think* we could just make the child process exit with failure and have that happen. This is doable.
3. Disconnect `processOutputs`, inside the `executeCdkCommandAndCleanupDirectory` function under test, from the real standard input and output streams of the process running the test, and connect it instead to some kind of streams that the test can manipulate, so that the test can watch for the forwarded MFA prompt and provide the fake MFA code. I *think* that can translate to something like "mock `fmt.Printf` and `fmt.Scanln`", if we're comfortable *requiring* that those *particular* functions be used as part of the contract of the function we're testing (or if we're just willing to change the test if we decide to use something else later). But I don't really have much of an idea of how to go about doing this.

It looks like the mocking stuff in use here already, for `os` for checking if the temporary directory removal method is called as expected, is actually calling [some generated code that lives in this repo](https://github.com/aws/amazon-genomics-cli/blob/29f8e9ee012e09ca294547ea3dba225304b36bfa/packages/cli/internal/pkg/mocks/io/mock_interfaces.go). There's [documentation here](https://github.com/aws/amazon-genomics-cli/blob/3d29cae963cd1a0afbbd6cddad42860b693a4c10/packages/cli/TESTING.md#mocks) about how this mock code is to be generated, from an "interface" file.

Actually, it looks like the [interface file](https://github.com/aws/amazon-genomics-cli/blob/29f8e9ee012e09ca294547ea3dba225304b36bfa/packages/cli/internal/pkg/mocks/io/interfaces.go) isn't just for `os` but actually covers a bunch of different system interfaces. Each of them has to be created and attached to the "controller" [like this](https://github.com/aws/amazon-genomics-cli/blob/a5338edad6e9b4cc9eebc0ed938e956cce15a17b/packages/cli/internal/pkg/aws/cdk/execute_cdk_test.go#L48-L49), presumably.

So I could add a `Fmt` type to that interface file, and rerun the mock generator to get a `MockFmt`, which I could then import and attach to the tests' controller. Then I could work out exactly what `fmt.Printf` and `fmt.Scanln` calls I expect to see, and what I would need them to return, and I could program the mock to expect and return those values for that test.

But I'm not entirely sure that's going to work. For no documented reason, the actual code is [using package-scope variables `osRemoveAll` and `execCommand`](https://github.com/aws/amazon-genomics-cli/blob/304a1787fe928e0b5832e535657594f73ad510f5/packages/cli/internal/pkg/aws/cdk/execute_cdk.go#L18-L21) instead of `os.RemoveAll` and `os/exec.Command`. The tests are, also for no documented reason, [swapping these out](https://github.com/aws/amazon-genomics-cli/blob/a5338edad6e9b4cc9eebc0ed938e956cce15a17b/packages/cli/internal/pkg/aws/cdk/execute_cdk_test.go#L50-L54). `osRemoveAll` is in particular swapped fro something from the `MockOS`.

Assuming that the tests are doing this for a good reason, that would mean that parts of the Go standard library that live on global objects, like `os.RemoveAll` and also `fmt.Printf`, cannot actually be mocked in the normal way and replaced out from under the code under test; the code has to be adjusted to cooperate and to allow test functions to be swapped in.

So that in turn would mean I can't really test this as currently written, using the mocking approach that the project usually uses; I'd have to go in and add a `fmtPrintf` and `fmtScanln` that could be swapped out by the test code.

And at that point, really I'd want to just redesign this a bit for better testability: abstract out the thing that is used to prompt for and receive the MFA code, and use a console-based one when running for real, and a special test one when testing.

What do you think is the best approach here, @nbraid? Am I right in my guesses about why the existing test code is doing what it is doing?

Reply:
I'll admit to being a Go novice, and as confused by a lot of what you called out as you are 🙂

If you have refactoring that you think ought to be done, then feel free.

I think what I mostly want to ensure is that changing from simply logging the lines to manually splitting runes and looking for newlines is working, as well as accepting some MFA input.

I think the existing tests somewhat cover the former, although none of the [test commands](https://github.com/aws/amazon-genomics-cli/blob/a5338edad6e9b4cc9eebc0ed938e956cce15a17b/packages/cli/internal/pkg/aws/cdk/execute_cdk_test.go#L155-L158) cover multiple lines at the moment.

For the latter, I think adding a test command that prints the MFA prompt like you suggested, and then verifying that we wait for and pass along input before proceeding would suffice?

Maybe this is harder than I'm realizing. I assume it's made more complex by the fact that we almost entirely capture stdout/stderr of the CDK command and end up piping it through Zerolog rather than using the Go primitives? It doesn't look like the code has any mechanism to pass in pipes to get access to the input/output of the command itself. Would something like that work? Maybe that's what you're referring to by "use a console-based one when running for real, and a special test one when testing."

Would something like

```go
func invokeCdk(commandArgs []string) {
	invokeCdkWithPipes(commandArgs,
		debugLogWriter(),
		progressBarWriter(),
		os.Stdin,
	)
}

func invokeCdkWithPipes(commandArgs []string, stdout io.Writer, stderr io.Writer, stdin io.Reader){
	cmd := execCommand("cdk", commandArgs...)
	cmd.Stdout = stdout
	cmd.Stderr = stderr
	cmd.Stdin = stdin
}
```

allow you to get access to the inputs/outputs from your test and verify them as you need to?

Reply:
I think the tests need to talk to the I/O of the main Go process, rather than the (fake) CDK process. I can probably swap out a couple of streams, I think; I don't want you to have to pass in streams to get MFA on to all the functions.